### PR TITLE
feat: New redis source

### PR DESCRIPTION
**Pull request by shumin1027**
a plugin to support redis(list、channel) as a data source

Context Issue: #5696 

the example config:

```toml
[sources.in]
type = "redis"
url = "redis://127.0.0.1:6379/0"
data_type = "list"
key = "vector"
list.method = "brpop"
redis_key = "redis_key"
```

### CODE REVIEW

**Comment by jszwedko** [comment]
@shumin1027 just noting that #7078 was merged 🎉 . I'm happy to review this one too once you've rebased on master.

**Comment by pachico** [comment]
I'm just expressing here my utmost interest for this feature :)

**Comment by shumin1027** [comment]
@jszwedko @pablosichert 
Just hope you can arrange some time to review this PR :)

**Comment by BirdyUK** [comment]
@shumin1027 is there any way i can build from the source you have already made that is awaiting the merge? Currently, i am pulling my hair out trying to get a high-performance pipeline into click house  [Vector works perfectly but we now need to add Redis as a test stream so I'm almost praying i can still use vector somehow!]

**Comment by spencergilbert** [comment]
> @shumin1027 is there any way i can build from the source you have already made that is awaiting the merge? Currently, i am pulling my hair out trying to get a high-performance pipeline into click house [Vector works perfectly but we now need to add Redis as a test stream so I'm almost praying i can still use vector somehow!]

@BirdyUK you should be able to check out the branch locally and use our make targets to build a custom binary, in the meantime I bumped this internally to see if we can get it reviewed and merged.

File: src/sources/redis/list.rs
Comment (pablosichert) [inline src/sources/redis/list.rs:111]: Not sure these helpers here are super useful, since they can be replaced with

```rust
conn.blpop(key, timeout).await.map(|(_, value)| value)
```

on the call site.

File: src/sources/redis/list.rs
Comment (pablosichert) [inline src/sources/redis/list.rs:98]: I think it would be nice to make the `timeout` argument configurable from the source config.

File: src/sources/redis/mod.rs
Comment (pablosichert) [inline src/sources/redis/mod.rs:99]:
```suggestion
    let client = build_client(config).map_err(|_| "Failed to open redis client.".into())?;
```

**Review by pablosichert — COMMENTED** [review]
Hey @shumin1027, thanks a lot for your contribution!

I have left a couple of suggestions for improvement below.

Also worth noting that it would be nice to integrate with `codecs::Decoder` in this new source. However, we can do that in a separate PR once we merge the initial version in.

File: src/sources/redis/list.rs
Comment (shumin1027) [reply src/sources/redis/list.rs:98]: The timeout argument  specifying the maximum number of seconds to block. A timeout of zero can be used to block indefinitely. I think this parameter does not need to be configured

Comment (shumin1027) [reply src/sources/redis/list.rs:98]: @pablosichert 
The timeout argument is specifying the maximum number of seconds to block. A timeout of zero can be used to block indefinitely. I think this parameter does not need to be configured

File: src/sources/redis/list.rs
Comment (shumin1027) [reply src/sources/redis/list.rs:94]: great 👍

File: src/sources/redis/list.rs
Comment (shumin1027) [reply src/sources/redis/list.rs:38]: @pablosichert 
thanks，I don’t know hao to replace `tokio::select! {}` whih  `and_then` and `or_else` , I also see that there is still a lot of code using this usage :

```rust
loop {
    tokio::select! {
    ……
    }
}
```

Is it possible to keep this usage and modify it later?  After all, it can work correctly at the moment

**Comment by shumin1027** [comment]
> @shumin1027 is there any way i can build from the source you have already made that is awaiting the merge? Currently, i am pulling my hair out trying to get a high-performance pipeline into click house [Vector works perfectly but we now need to add Redis as a test stream so I'm almost praying i can still use vector somehow!]

@BirdyUK 
You can look about this branches：
https://github.com/shumin1027/vector/tree/redis-source

**Comment by BirdyUK** [comment]
> 
> 
> > @shumin1027 is there any way i can build from the source you have already made that is awaiting the merge? Currently, i am pulling my hair out trying to get a high-performance pipeline into click house [Vector works perfectly but we now need to add Redis as a test stream so I'm almost praying i can still use vector somehow!]
> 
> @BirdyUK You can look about this branches： https://github.com/shumin1027/vector/tree/redis-source

Thanks, I was actually able to get it to work and read from the redis channel works great.

The only issue i have now which unfortunately prevents me from being able to use Vector as intended is the clickhouse plugin does not contain the recent vector clickhouse changes "skip_unknown_fields", I am hoping that skip_unknown_fields will work once merged as it appears skip_unknown_fields was implemented while your PR has been waiting approval.

Side note: Do you also think you would be able to implement Redis streams as a source in the near future?

Thanks :-)

**Comment by shumin1027** [comment]
> Thanks, I was actually able to get it to work and read from the redis channel works great.
> 
> The only issue i have now which unfortunately prevents me from being able to use Vector as intended is the clickhouse plugin does not contain the recent vector clickhouse changes "skip_unknown_fields", I am hoping that skip_unknown_fields will work once merged as it appears skip_unknown_fields was implemented while your PR has been waiting approval.
> 
> Side note: Do you also think you would be able to implement Redis streams as a source in the near future?
> 
@BirdyUK 
Thanks for your test,  support for `Redis Stream` is a great feature, I will try to implement it

**Comment by BirdyUK** [comment]
> @BirdyUK Thanks for your test, support for `Redis Stream` is a great feature, I will try to implement it

I just noticed your recent branch now includes Vector's recently added clickhouse feature "skip_unknown_fields"... I have cloned the branch and done a "make build", i can report back that everything works perfectly including clickhouse 🥇 

I would just like to say a huge thank you for putting the time and effort into helping the awesome community, you have done an amazing job. I know many of us Redis users will thank you for many moons to come :-)

The cherry on the top would be Redis Stream, potentially a config option to XACK and XDEL the message once Vector has read the message?

Again, thanks for the amazing contributions it has been a huge help.

**Comment by shumin1027** [comment]
> > @BirdyUK Thanks for your test, support for `Redis Stream` is a great feature, I will try to implement it
> 
> I just noticed your recent branch now includes Vector's recently added clickhouse feature "skip_unknown_fields"... I have cloned the branch and done a "make build", i can report back that everything works perfectly including clickhouse 1st_place_medal
> 
> I would just like to say a huge thank you for putting the time and effort into helping the awesome community, you have done an amazing job. I know many of us Redis users will thank you for many moons to come :-)
> 
> The cherry on the top would be Redis Stream, potentially a config option to XACK and XTRIM the message once Vector has read the message?
> 
> Again, thanks for the amazing contributions it has been a huge help.

@BirdyUK
After this PR be merged，we can open a new issue for `Redis Stream`, such as: #10042

**Comment by BirdyUK** [comment]
@shumin1027 - Could you please advise if the following is expected behavior or if something is not right:

Vector is connected to Redis channel and reading data perfectly fine, however, if the service  sending data to redis happens to stop for X amount of time and then restart the reconnect to redis... vector no longer picks up the new messages going back onto the same redis channel?

However, if i keep open a terminal window to redis-cli and subscribe to the same channel vector is subscribed to then the terminal will still detect the new messages regardless of how long the producer has been inactive.

Does that make sense?

It appears that i need to restart vector if i restart the redis producer otherwise vector will not read from the channel which seems odd because in theory they are decoupled? I am thinking that If redis was the issue then the CLI would not detect the messages either?

Thanks

**Comment by jszwedko** [comment]
Hi @shumin1027 ! Apologies for the long delay in review on this. Just a note that I plan to review this again this week.

We are trying to improve our PR review processes to avoid this in the future.

**Review by jszwedko — COMMENTED** [review]
Thanks for this contribution @shumin1027 !

Rather than leaving a bunch of detailed review comments, I just made the changes since I figured it'd be easier to show than try to explain. Let me know what you think or if you disagree with any of them.

I'll tag another team member to review since I made significant changes here, but we should be able to get this merged soon.

**Review by bruceg — CHANGES_REQUESTED** [review]
This looks reasonably straighforward. I left comments about a couple of minor issues below.

File: docs/reference/components/sources/redis.cue
Comment (jszwedko) [reply docs/reference/components/sources/redis.cue:6]: Oh yes, thanks!

File: docs/reference/components/sources/redis.cue
Comment (jszwedko) [reply docs/reference/components/sources/redis.cue]: Good call, I fixed and clarified in [88c6a34](https://github.com/vectordotdev/vector/pull/7096/commits/88c6a3411cc2f6d1d67d2b9165182203a43dc46b)

File: docs/reference/components/sources/redis.cue
Comment (jszwedko) [reply docs/reference/components/sources/redis.cue:6]: [88c6a34](https://github.com/vectordotdev/vector/pull/7096/commits/88c6a3411cc2f6d1d67d2b9165182203a43dc46b)

**Review by bruceg — APPROVED** [review]

**Comment by shumin1027** [comment]
> Thanks for this contribution @shumin1027 !
> 
> Rather than leaving a bunch of detailed review comments, I just made the changes since I figured it'd be easier to show than try to explain. Let me know what you think or if you disagree with any of them.
> 
> I'll tag another team member to review since I made significant changes here, but we should be able to get this merged soon.

Thank you @jszwedko , I'm sorry for causing a lot of unnecessary trouble to your work

**Comment by jszwedko** [comment]
> > Thanks for this contribution @shumin1027 !
> > Rather than leaving a bunch of detailed review comments, I just made the changes since I figured it'd be easier to show than try to explain. Let me know what you think or if you disagree with any of them.
> > I'll tag another team member to review since I made significant changes here, but we should be able to get this merged soon.
> 
> Thank you @jszwedko , I'm sorry for causing a lot of unnecessary trouble to your work

No worries! I appreciate you starting the work.

**Pull request merged**
